smart_card_connector_app/src/reader-tracker.js

```diff
@@ -0,0 +1,212 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview This file contains the class used to track readers and their
+ * status based on messages received from the hook setup on PC/SC code inside
+ * NaCl (check readerfactory_nacl.cc for more information).
+ */
+
+goog.provide('GoogleSmartCard.ReaderTracker');
+
+goog.require('GoogleSmartCard.DebugDump');
+goog.require('GoogleSmartCard.Logging');
+goog.require('GoogleSmartCard.TypedMessage');
+goog.require('goog.asserts');
+goog.require('goog.log.Logger');
+goog.require('goog.messaging.AbstractChannel');
+goog.require('goog.object');
+goog.require('goog.structs.Map');
+
+goog.scope(function() {
+
+/** @const */
+var READER_TRACKER_LOGGER_TITLE = 'ReaderTracker';
+
+/** @const */
+var GSC = GoogleSmartCard;
+
+/**
+ * Enum for possible values of ReaderInfo status.
+ * @enum {string}
+ */
+var ReaderStatus = {
+  INIT: 'init',
+  SUCCESS: 'success',
+  FAILURE: 'failure'
+};
+
+/**
+ * Structure used to store information about the reader.
+ * @param {string} name
+ * @param {ReaderStatus} status
+ * @param {string=} opt_error
+ * @constructor
+ */
+var ReaderInfo = function(name, status, opt_error) {
+  this['name'] = this.name = name;
+  this['status'] = this.status = status;
+  this['error'] = this.error = opt_error;
+};
+
+/**
+ * This class tracks readers, provides methods to retrieve the list of readers
+ * and get updates on change, and listens to messages received from the hook
+ * living in the NaCl space.
+ * @param {!goog.messaging.AbstractChannel} messageChannel
+ * @param {!goog.log.Logger} parentLogger
+ * @constructor
+ */
+GSC.ReaderTracker = function(messageChannel, parentLogger) {
+  /** @private */
+  this.logger_ = GSC.Logging.getChildLogger(
+      parentLogger, READER_TRACKER_LOGGER_TITLE);
+
+  messageChannel.registerService(
+      'reader_init_add', this.readerInitAddListener_.bind(this), true);
+  messageChannel.registerService(
+      'reader_finish_add', this.readerFinishAddListener_.bind(this), true);
+  messageChannel.registerService(
+      'reader_remove', this.readerRemoveListener_.bind(this), true);
+
+  /**
+   * @type {!goog.structs.Map.<number, !ReaderInfo>}
+   * @private
+   */
+  this.readers_ = new goog.structs.Map;
+
+  /**
+   * @type {!Array.<!function(!Array.<!ReaderInfo>)>}
+   * @private
+   */
+  this.onUpdateListeners_ = [];
+
+  this.logger_.fine('Initialized');
+};
+
+/** @const */
+var ReaderTracker = GSC.ReaderTracker;
+
+/**
+ * @param {!Object|string} message
+ * @private
+ */
+ReaderTracker.prototype.readerInitAddListener_ = function(message) {
+  /** @type {string} */
+  var name = message['readerName'];
+  /** @type {number} */
+  var port = message['port'];
+  /** @type {string} */
+  var device = message['device'];
+
+  this.logger_.info(
+      'readerInitAddListener_ called for ' + name + ' (port ' + port +
```

Comment:
nit: Instead of including the private member names (like "readerInitAddListener_" in this case) into the log messages, better print some human-readable information (like "Adding a new reader "<name>" with port <port>, device "<device>" and status "initializing"" in this case).